core/src/main/java/org/jdbi/v3/core/statement/OutParameters.java

```diff
@@ -18,339 +18,540 @@
-            }
-            return null;
-        }
-
-        if (obj instanceof Number) {
-            return (Number) obj;
-        } else {
-            throw new IllegalArgumentException(format("Parameter %d is a %s, not a number", pos, obj.getClass()));
+        try {
+            return getObject(pos, Number.class);
+        } catch (ClassCastException e) {
+            throw new IllegalArgumentException(format("Parameter #%d is not a number", pos), e);
         }
     }
 
     @Nullable
     @SuppressWarnings("JavaUtilDate")
     private Long getEpoch(String name) {
-        Object obj = map.get(name);
-
-        if (obj == null) {
-            if (!map.containsKey(name)) {
-                throw new IllegalArgumentException(format("Parameter '%s' does not exist", name));
-            }
-
-            return null;
-        }
-
-        if (obj instanceof java.util.Date) {
-            return ((java.util.Date) obj).getTime();
-        } else {
-            throw new IllegalArgumentException(format("Parameter '%s' is a %s, not a Date", name, obj.getClass()));
+        try {
+            var date = getObject(name, java.util.Date.class);
+            return date == null ? null : date.getTime();
+        } catch (ClassCastException e) {
+            throw new IllegalArgumentException(format("Parameter '%s' is not a java.util.Date", name), e);
         }
     }
 
     @Nullable
     @SuppressWarnings("JavaUtilDate")
     private Long getEpoch(int pos) {
-        Object obj = map.get(pos);
-        if (obj == null) {
-            if (!map.containsKey(pos)) {
-                throw new IllegalArgumentException(format("Parameter at %d does not exist", pos));
-            }
+        try {
+            var date = getObject(pos, java.util.Date.class);
+            return date == null ? null : date.getTime();
+        } catch (ClassCastException e) {
+            throw new IllegalArgumentException(format("Parameter #%d is not a java.util.Date", pos), e);
+        }
+    }
+
+    void putValueSupplier(int index, String key, Supplier<Object> value) {
+        map.putValueSupplier(index, key, value);
+    }
 
-            return null;
+    private static final class ParameterValueMap {
+        private final Map<Object, Supplier<Object>> map = new HashMap<>();
+
+        void putValueSupplier(int index, String key, Supplier<Object> value) {
+            map.put(index, value);
+            if (key != null) {
+                map.put(key, value);
+            }
         }
 
-        if (obj instanceof java.util.Date) {
-            return ((java.util.Date) obj).getTime();
-        } else {
-            throw new IllegalArgumentException(format("Parameter at %d is a %s, not a Date", pos, obj.getClass()));
+        Object getValue(int index) {
+            if (!map.containsKey(index)) {
+                throw new IllegalArgumentException(format("Parameter #%d does not exist", index));
+            }
+            Supplier<Object> supplier = map.get(index);
```

Comment:
It is trivially more efficient to call the `get(index)` first, then only check `containsKey` after observing the `supplier != null` which is computed anyway.
(The old code was careful to do this. I am not claiming that this is a big deal.)

Reply:
I am not disagreeing but the old code was really hard to understand (I did not understand it. :-) ).  I simplified it.

Reply:
Maybe the code reads a tiny bit simpler, but now every time this functionality is used we will double the number of map looks, for every user, forever :(

The code was written the way it was previously quite intentionally.

Reply:
Oh, I see, it is much simpler now. Thanks for fixing this.